velox/exec/tests/VectorHasherTest.cpp

```diff
@@ -693,3 +694,89 @@ TEST_F(VectorHasherTest, computeValueIdsStrings) {
+  // We sample with large but not end-reaching ranges of different ints.
+  // We check that ranges are rounded up to ends of the data types.
+  // We check that nulls, min and max values in each column combine to make
+  // distinct normalized keys.
+  constexpr int64_t kTinyRange = std::numeric_limits<uint8_t>::max();
+  constexpr int64_t kSmallRange = std::numeric_limits<uint16_t>::max();
+  constexpr int64_t kIntRange = std::numeric_limits<uint32_t>::max();
+
+  using exec::VectorHasher;
+  auto tinySample = vectorMaker_->flatVector(std::vector<int8_t>{-100, 110});
+  auto smallSample =
+      vectorMaker_->flatVector(std::vector<int16_t>{-30000, 31000});
+  auto intSample =
+      vectorMaker_->flatVector(std::vector<int32_t>{-2001000000, 2000000000});
+  auto tinyHasher = VectorHasher::create(TINYINT(), 0);
+  auto smallHasher = VectorHasher::create(SMALLINT(), 1);
+  auto intHasher = VectorHasher::create(INTEGER(), 2);
+  raw_vector<uint64_t> result;
+  SelectivityVector rows(2);
+  tinyHasher->computeValueIds(*tinySample, rows, result);
+  smallHasher->computeValueIds(*smallSample, rows, result);
+  intHasher->computeValueIds(*intSample, rows, result);
+  uint64_t distinct;
+  uint64_t range;
+  tinyHasher->cardinality(50, range, distinct);
+  // Distinct count is 4 because 2 values + 50% is 3 and 1 for the null value.
+  EXPECT_EQ(4, distinct);
+  EXPECT_EQ(range, kTinyRange + 2);
+  smallHasher->cardinality(50, range, distinct);
+  EXPECT_EQ(4, distinct);
+  EXPECT_EQ(range, kSmallRange + 2);
+  intHasher->cardinality(50, range, distinct);
+  EXPECT_EQ(4, distinct);
+  EXPECT_EQ(range, kIntRange + 2);
+  auto multiplier1 = tinyHasher->enableValueRange(1, 50);
+  auto multiplier2 = smallHasher->enableValueRange(multiplier1, 50);
+  auto multiplier3 = intHasher->enableValueRange(multiplier2, 50);
+  // The total combinations are the product of the ranges + 1 for null.
+  EXPECT_EQ(
+      (kTinyRange + 2) * (kSmallRange + 2) * (kIntRange + 2), multiplier3);
+  auto tinyData =
+      vectorMaker_->flatVector<int8_t>(0, [](auto row) { return row; });
+  auto smallData =
+      vectorMaker_->flatVector<int16_t>(0, [](auto row) { return row; });
+  auto intData =
+      vectorMaker_->flatVector<int32_t>(0, [](auto row) { return row; });
+  for (auto tinyId = 0; tinyId < 4; ++tinyId) {
+    for (auto smallId = 0; smallId < 4; ++smallId) {
+      for (auto intId = 0; intId < 4; ++intId) {
+        append<int8_t>(tinyData, tinyId);
+        append<int16_t>(smallData, smallId);
+        append<int32_t>(intData, intId);
+      }
+    }
+  }
+  result.resize(tinyData->size());
+  rows.resize(tinyData->size());
+  tinyHasher->computeValueIds(*tinyData, rows, result);
+  smallHasher->computeValueIds(*smallData, rows, result);
+  intHasher->computeValueIds(*intData, rows, result);
+  std::unordered_set<uint64_t> uniques;
+  for (auto id : result) {
+    uniques.insert(id);
+  }
+  EXPECT_EQ(tinyData->size(), uniques.size());
```

Comment:
What does this test check? Would `result` be overwritten at the lines of tinyHasher/smallHasher/intHasher->computeValueIds(...)?